### PR TITLE
fix some bugs in like Expr

### LINKED ISSUE #46592: backslash in `LIKE` is potentially not handled correctly
see https://fiddle.clickhouse.com/55bd387b-83df-49a2-bab3-866e8e6457e5

```sql
CREATE TABLE t (p String) ENGINE=Memory;

insert into t values ('c:\\windows\\deyk.exe');
insert into t values ('C:\\Windows\\DEyk.exe');
insert into t values ('c:\\windows\\explorer.exe');
insert into t values ('C:\\Windows\\System32\\cmd.exe');
insert into t values ('C:\\Windows\\System32\\conhost.exe');
insert into t values ('C:\\Windows\\System32\\findstr.exe');
insert into t values ('c:\\windows\\system32\\net.exe');
insert into t values ('c:\\windows\\system32\\net1.exe');
insert into t values ('C:\\Windows\\System32\\svchost.exe');
insert into t values ('C:\\Windows\\System32\\wbem\\WMIC.exe');
insert into t values ('c:\\windows\\system32\\wermgr.exe');
insert into t values ('c:\\windows\\syswow64\\wbem\\wmic.exe');

SELECT '0',* FROM t FORMAT TabSeparatedRaw;

SELECT '1',* FROM t WHERE `p` LIKE '%Windows\System32%'; -- expect non-empty result
SELECT '2',* FROM t WHERE `p` LIKE '%Windows\\System32%'; -- expect non-empty result
SELECT '3',* FROM t WHERE `p` LIKE '%Windows\\\System32%'; -- expect empty result
SELECT '4',* FROM t WHERE `p` LIKE '%Windows\\\\System32%'; -- expect empty result
SELECT '5',* FROM t WHERE `p` LIKE '%Windows\\\\\System32%'; -- expect empty result
```

### LINKED ISSUE #46654: % in LIKE is potentially not handled correctly (with line break)
LIKE expr behaves very very weird when meating \n and \\

https://fiddle.clickhouse.com/54c8d7c6-7201-4794-8609-f9c6125de079

```sql
CREATE TABLE MY_TABLE(xForwardedFor String) ENGINE=Memory;

INSERT INTO MY_TABLE(xForwardedFor) VALUES ( 'X-Forwarded-For: (null)\nCdn-Src-Ip: (null)\nClientip: (null)');

SELECT count(*)
FROM MY_TABLE
WHERE like(toString(`xForwardedFor`), '%X\\-Forwarded\\-For%');

SELECT count(*)
FROM MY_TABLE
WHERE like(toString(`xForwardedFor`), '%X\\-Forwarded\\-For%Cdn%');

SELECT count(*)
FROM MY_TABLE
WHERE like(toString(`xForwardedFor`), '%X-Forwarded-For%');

SELECT count(*)
FROM MY_TABLE
WHERE like(toString(`xForwardedFor`), '%X-Forwarded-For%Cdn%');

SELECT count(*)
FROM MY_TABLE
WHERE like(toString(`xForwardedFor`), '%X\\-Forwarded\\-For: \\(null\\)%\\\nCdn\\-Src\\-Ip: \\(null\\)%\\\nClientip: \\(null\\)%');

```

response

```
1
0
1
1
0
```

**Does it reproduce on recent release?**

yes

**Expected behavior**

response

```
1
1
1
1
1
```
**Additional context**

### PR DESCRIPTION
**Pull request by XenoAmess**
close https://github.com/ClickHouse/ClickHouse/issues/46592

close https://github.com/ClickHouse/ClickHouse/issues/46654

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Fix some bugs in like Expr, especially part dealing with / escape.

It shall have some small performance benifit like dealing with something like "%%%%%%%%%%%%%%%%%%%%", and some minor performance benifit for better iteration usage, but that is not the main point.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by pufit — CHANGES_REQUESTED** [review]

**Comment by pufit** [comment]
Also, it will be nice to have some tests for it

File: src/Functions/likePatternToRegexp.h [hunk 1]
```diff
@@ -20,53 +20,64 @@ inline String likePatternToRegexp(std::string_view pattern)
...
+    char char_slot;
     if (pos < end && *pos == '%')
-        ++pos;
+    {
+        while (++pos < end && *pos == '%')
+        {
+        }
+    }
     else
+    {
```
[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/Functions/likePatternToRegexp.h:31]: according to rule 17 it is optional, so there might be no necessary to remove it

File: src/Functions/likePatternToRegexp.h [hunk 2]
```diff
@@ -20,53 +20,64 @@ inline String likePatternToRegexp(std::string_view pattern)
...
+                }
                 break;
             case '_':
                 res += ".";
                 break;
             case '\\':
-                if (pos + 1 == end)
-                    throw Exception(ErrorCodes::CANNOT_PARSE_ESCAPE_SEQUENCE, "Invalid escape sequence at the end of LIKE pattern");
-                /// Known escape sequences.
-                if (pos[1] == '%' || pos[1] == '_')
```
[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/Functions/likePatternToRegexp.h:49]: > I believe we shouldn't remove this part because "%" should be transformed into "%".

no need to, as according to regex's rule, `/%` = `%`

File: src/Functions/likePatternToRegexp.h [hunk 3]
```diff
@@ -20,53 +20,64 @@ inline String likePatternToRegexp(std::string_view pattern)
...
     else
+    {
         res = "^";
-
-    while (pos < end)
+    }
+    while (pos != end)
     {
-        switch (*pos)
+        char_slot = *(pos++);
```
[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/Functions/likePatternToRegexp.h:36]: > Do you think it's necessary? Because it looks much less intuitive to me. I understand it could probably be faster, but pattern strings are usually not so big. So it is reasonable to sacrifice readability in this case.

well it could be changed to have better readability instead of performance.

**Comment by XenoAmess** [comment]
> Also, it will be nice to have some tests for it

need some help with this.

where should I add test cases in this repo? please give an example, I would add some tests.

Also, don't understand why FastTest failed... please guide me how to see the failing reason, thanks.

File: src/Functions/likePatternToRegexp.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/Functions/likePatternToRegexp.h]: @pufit done.

**Comment by pufit** [comment]
> > Also, it will be nice to have some tests for it
> 
> @pufit
> 
> need some help with this.
> 
> where should I add test cases in this repo? please give an example, I would add some tests.
> 

Here you can find out test guides. https://clickhouse.com/docs/en/development/adding_test_queries
TL;DR You need to go here https://github.com/ClickHouse/ClickHouse/tree/master/tests/queries/0_stateless and add a new script (.sql or .sh) and .reference file with its supposed output.

> Also, don't understand why FastTest failed... please guide me how to see the failing reason, thanks.
It seems that some of that queries didn't finish correctly https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01847_bad_like.sql#L11

File: src/Functions/likePatternToRegexp.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (pufit) [reply src/Functions/likePatternToRegexp.h:30]: Yeah, my bad. Disregard

File: src/Functions/likePatternToRegexp.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (pufit) [reply src/Functions/likePatternToRegexp.h:49]: We are using `re2` for parsing regex. And as far as I understand, `/%` != `%` in it (at least in online web parser)

**Comment by pufit** [comment]
> > Also, don't understand why FastTest failed... please guide me how to see the failing reason, thanks.
> 
> It seems that some of that queries didn't finish correctly https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01847_bad_like.sql#L11

And you can see the failure message if you click the `details` button for failed CI check and search for `[ FAIL ]

File: src/Functions/likePatternToRegexp.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/Functions/likePatternToRegexp.h:49]: > We are using `re2` for parsing regex. And as far as I understand, `/%` != `%` in it (at least in online web parser)

you are correct. fixing.

[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/Functions/likePatternToRegexp.h:49]: @pufit done

File: src/Functions/likePatternToRegexp.h [hunk 4]
```diff
@@ -21,52 +21,84 @@ inline String likePatternToRegexp(std::string_view pattern)
...
+                {
                     res += ".*";
+                }
                 else
+                {
                     return res;
+                }
                 break;
             case '_':
+                ++pos;
```
Comment (rschu1ze) [inline src/Functions/likePatternToRegexp.h:64]: For consistency accross the `case`s, could `++pos` be deferred as much as possible? E.g. here, it could be done immediately before `break`.

File: src/Functions/likePatternToRegexp.h [hunk 5]
```diff
@@ -21,52 +21,84 @@ inline String likePatternToRegexp(std::string_view pattern)
...
+            case '+':
+            case '{':
                 res += '\\';
                 res += *pos;
+                ++pos;
                 break;
             case '%':
-                if (pos + 1 != end)
+                ++pos;
+                if (pos != end)
```
Comment (rschu1ze) [inline src/Functions/likePatternToRegexp.h:54]: Braces around single-line if-statements and the like (e.g. l. 70) look quite unusual in the context of the ClickHouse codebase. The next person working on the code will likely remove them. Better don't even add them in the first place :)

**Comment by rschu1ze** [comment]
Question: Would you like to integrate my doc changes #46722 into this PR and document the escaping rules for function `LIKE`? Basically replace sentence `See the note on escaping in the description of the match function.` for that. As far as I read, there was some confusion around them in the two bugs referenced by this PR, so we better document them once and for all.

**Comment by XenoAmess** [comment]
> Question: Would you like to integrate my doc changes #46722 into this PR and document the escaping rules for function `LIKE`? Basically replace sentence `See the note on escaping in the description of the match function.` for that. As far as I read, there was some confusion around them in the two bugs referenced by this PR, so we better document them once and for all.

I'd like to but as this pr changes some behavior of LIKE function, I think it better wait for some code review first, and after the codes (and new behaviors) accepted then we dealing with documents part...

File: src/Functions/likePatternToRegexp.h (hunk 4 above)
Comment (XenoAmess) [reply src/Functions/likePatternToRegexp.h]: @rschu1ze done.

File: src/Functions/likePatternToRegexp.h (hunk 5 above)
Comment (XenoAmess) [reply src/Functions/likePatternToRegexp.h:48]: Interesting. 
As both @pufit and @rschu1ze requested for this, I think I'd better follow the current codebase's style as well

File: tests/queries/0_stateless/01847_bad_like.reference [hunk 6]
```diff
@@ -1,15 +1,31 @@
 1
+0
```
Comment (pufit) [inline tests/queries/0_stateless/01847_bad_like.reference:2]: Why is this zero?
`SELECT '\w' LIKE '\w%';`

Comment (XenoAmess) [reply tests/queries/0_stateless/01847_bad_like.reference:2]: because `'\w%'` should be regex `^w.*` , then that is it.

Comment (XenoAmess) [reply tests/queries/0_stateless/01847_bad_like.reference:2]: @pufit see example in https://github.com/ClickHouse/ClickHouse/issues/46682

we want to make sure wether we enter like->regex way, or like->substr way, they behaves samely.

d in like->substr way \w is handled as w.and in like->substr way \w is handled as w.

Comment (XenoAmess) [reply tests/queries/0_stateless/01847_bad_like.reference:2]: maybe a better demo is https://fiddle.clickhouse.com/5cd99b64-dc92-4041-802c-9fbef1589af2

```
SELECT '\awwww' LIKE '%\wwww%';
```

= 1

but

```
SELECT '\awwww' LIKE '%\ww%w%';
```

= 0


because  `'%\wwww%'` enters like->substr way, and it becomes `substring wwww`

but `'%\ww%w%'` enters regex->substr way. to make logic similar, it should become something like `.*ww.*w.*`, but actually the \ behaves differently than like->substr way

Comment (pufit) [reply tests/queries/0_stateless/01847_bad_like.reference:2]: Hi! Sorry for the long response. For compatibility reasons, we want to interpret unknown escape sequences with keeping the backslash character '\w' as \w , not w . Therefore, this test should return 1. Otherwise, it can have a significant impact on ClickHouse users.

Comment (XenoAmess) [reply tests/queries/0_stateless/01847_bad_like.reference:2]: > Hi! Sorry for the long response. For compatibility reasons, we want to interpret unknown escape sequences with keeping the backslash character '\w' as \w , not w . Therefore, this test should return 1. Otherwise, it can have a significant impact on ClickHouse users.

Oh, then could you mind me change the like -> substring way to '\w' as \w , to make things consistence？

**Comment by rschu1ze** [comment]
All necessary fixes are integrated with https://github.com/ClickHouse/ClickHouse/pull/46833, https://github.com/ClickHouse/ClickHouse/pull/46867, https://github.com/ClickHouse/ClickHouse/pull/46858, https://github.com/ClickHouse/ClickHouse/pull/46869 and https://github.com/ClickHouse/ClickHouse/pull/46875. Sorry, but I think this PR is no longer needed. Thanks a ton for the super useful investigations anyways.
